src/util/code_observer.h

```diff
@@ -0,0 +1,161 @@
+/// table.  These classes essentially implement the observable pattern.
+///
+/// In general, you first register one or more observers with either
+/// code_observert or const_code_observert. The observer is a callback that is
+/// associated with a specific irep_idt. This is used to determine for which
+/// codet's the observer will be notified. The second step involves calling
+/// visit_symbols which will in turn iterate through every codet inside the
+/// given symbol-table and notify registered observers only if the id of the
+/// codet in question matches the associated id of the observer.
+///
+/// Example: Replacing all codet(ID_function_call) with codet(ID_skip)
+///
+/// \code
+///
+/// void A::replace_function_calls(codet &code)
+/// {
+///   PRECONDITION(code.get_statement()==ID_function_call);
+///   code = codet(ID_skip);
+/// }
+///
+/// ...
+///
+/// code_observert code_observer;
+/// code_observer.registerObserver(ID_function_call,
+///     &A::replace_function_calls,
+///     this,
+///     std::placeholders::_1);
+/// code_observer.visit_symbols(symbol_table);
+///
+/// ...
+///
+/// \endcode
+///
+/// Note: if the callback only needs to read (and not modify) the codet, then
+/// const_code_observert should be used otherwise use code_observert.
+///
+/// Note': each unique irep_idt can only be associated with one observer.
+
+#ifndef CPROVER_UTIL_CODE_OBSERVER_H
+#define CPROVER_UTIL_CODE_OBSERVER_H
+
+#include <map>
+#include <functional>
+#include <type_traits>
+
+#include "symbol_table.h"
+#include "std_code.h"
+
+/// \brief base class for all code observer implementations
+template<class T> class code_observer_baset
+{
+static_assert(std::is_base_of<codet, T>::value, "T is not derived from codet");
+public:
+  virtual ~code_observer_baset() {}
+
+  /// registers an observer, that is associated with specific codet id.
+  /// \param id: id of the codet to be registered
+  /// \param f: observer callback, that is associated with the given, \p id
+  /// \param args: arguments associated with the given callback, \p f
+  template<class F, class ...Args>
+    void registerObserver(const irep_idt &id, const F &&f, Args&&... args)
+    {
+      auto observer = std::bind(f, args ...);
+      code_observers_map.insert({id, observer});
+    }
+protected:
+  /// notifies (if any) the observer that is registered to receive a callback
+  /// for the given codet, \p code
+  /// \param code: will be passed to the registered observer that
+  /// that has the same associated id
+  void notify(T &code) const
+  {
+    auto it = code_observers_map.find(code.get_statement());
+    if(it != code_observers_map.end())
+      it->second(code);
+  }
+private:
+  using observerst = std::function<void(T&)>;
+  using code_observer_mapt = std::map<const irep_idt, const observerst>;
+  code_observer_mapt code_observers_map;
+};
+
+/// \brief codet observer, uses expression visitor to visit all the codet's
```

Comment:
This comment provides an implementation detail rather than focusing on the big picture. Write documentation by working backwards from the potential user, not for those editing the code.

Reply:
done